servant/server/src/Mu/Servant/Server.hs

```diff
@@ -0,0 +1,351 @@
+{-# language DataKinds             #-}
+{-# language DeriveGeneric         #-}
+{-# language FlexibleContexts      #-}
+{-# language FlexibleInstances     #-}
+{-# language GADTs                 #-}
+{-# language MultiParamTypeClasses #-}
+{-# language PolyKinds             #-}
+{-# language RankNTypes            #-}
+{-# language ScopedTypeVariables   #-}
+{-# language TypeApplications      #-}
+{-# language TypeFamilies          #-}
+{-# language TypeOperators         #-}
+{-# language UndecidableInstances  #-}
+{-# OPTIONS_GHC -Wall #-}
+
+module Mu.Servant.Server where
```

Comment:
It would also be great if you could add the information in https://github.com/higherkindness/mu-haskell/pull/217#issuecomment-696128262 as part of the module documentation, so we do not forget.

Reply:
Gladly. Ideally, I would like to use a custom `TypeError` to provide a nice error message when the user doesn't provide a certain instance. Currently, the error message looks like
```
/home/andre/mu-haskell/servant/server/exe/ExampleServer.hs:19:19: error:                                                                                                          • No instance for (HasServer                                                                                                                                                                       @{*}
                         (Mu.Servant.Server.PrefixRoute
                            (Mu.Servant.Server.Concat
                               @ghc-prim-0.5.3:GHC.Types.Symbol
                               (Mu.Servant.Server.FromMaybe
                                  @[ghc-prim-0.5.3:GHC.Types.Symbol]
                                  ((':)
                                     @ghc-prim-0.5.3:GHC.Types.Symbol
                                     "Greeter"
                                     ('[] @ghc-prim-0.5.3:GHC.Types.Symbol))
                                  (GetServiceAnnotationMay
                                     @ServantRoute
                                     @ghc-prim-0.5.3:GHC.Types.Symbol
                                     @ghc-prim-0.5.3:GHC.Types.Symbol
                                     @ghc-prim-0.5.3:GHC.Types.Symbol
                                     (AnnotatedPackage
                                        @ghc-prim-0.5.3:GHC.Types.Symbol
                                        @ghc-prim-0.5.3:GHC.Types.Symbol
                                        @ghc-prim-0.5.3:GHC.Types.Symbol
                                        @(Mu.Rpc.TypeRef ghc-prim-0.5.3:GHC.Types.Symbol)
                                        ServantRoute
                                        ('Mu.Rpc.Package
                                           @ghc-prim-0.5.3:GHC.Types.Symbol                                                                                                                                              @ghc-prim-0.5.3:GHC.Types.Symbol
                                           @ghc-prim-0.5.3:GHC.Types.Symbol                                                                                                                                              @(Mu.Rpc.TypeRef ghc-prim-0.5.3:GHC.Types.Symbol)
                                           ('Just @ghc-prim-0.5.3:GHC.Types.Symbol "helloworld")
                                           ((':)
                                              @(Mu.Rpc.Service
                                                  ghc-prim-0.5.3:GHC.Types.Symbol
                                                  ghc-prim-0.5.3:GHC.Types.Symbol
                                                  ghc-prim-0.5.3:GHC.Types.Symbol
                                                  (Mu.Rpc.TypeRef ghc-prim-0.5.3:GHC.Types.Symbol))
                                              ('Mu.Rpc.Service                                                                                                                                                                 @ghc-prim-0.5.3:GHC.Types.Symbol
                                                 @ghc-prim-0.5.3:GHC.Types.Symbol
                                                 @ghc-prim-0.5.3:GHC.Types.Symbol
                                                 @(Mu.Rpc.TypeRef ghc-prim-0.5.3:GHC.Types.Symbol)
                                                 "Greeter"
                                                 ((':)
                                                    @(Mu.Rpc.Method
                                                        @{*}
                                                        ghc-prim-0.5.3:GHC.Types.Symbol
                                                        ghc-prim-0.5.3:GHC.Types.Symbol
                                                        ghc-prim-0.5.3:GHC.Types.Symbol
                                                        (Mu.Rpc.TypeRef
                                                           ghc-prim-0.5.3:GHC.Types.Symbol))
                                                    ('Mu.Rpc.Method
                                                       @{*}
                                                       @ghc-prim-0.5.3:GHC.Types.Symbol
                                                       @ghc-prim-0.5.3:GHC.Types.Symbol
                                                       @ghc-prim-0.5.3:GHC.Types.Symbol
                                                       @(Mu.Rpc.TypeRef
                                                           ghc-prim-0.5.3:GHC.Types.Symbol)
                                                       "SayHello"
                                                       ((':)
                                                          @(Mu.Rpc.Argument
                                                              @{*}
                                                              ghc-prim-0.5.3:GHC.Types.Symbol
                                                              ghc-prim-0.5.3:GHC.Types.Symbol
                                                              (Mu.Rpc.TypeRef
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol))
                                                          ('Mu.Rpc.ArgSingle
                                                             @{*}
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             @(Mu.Rpc.TypeRef
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol)
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             ('Nothing
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol)
                                                             ('Mu.Rpc.SchemaRef
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                QuickstartSchema
                                                                "HelloRequest"))
                                                          ('[]
                                                             @(Mu.Rpc.Argument
                                                                 @{*}
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol
                                                                 (Mu.Rpc.TypeRef
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol))))
                                                       ('Mu.Rpc.RetSingle
                                                          @{*}
                                                          @(Mu.Rpc.TypeRef
                                                              ghc-prim-0.5.3:GHC.Types.Symbol)
                                                          @ghc-prim-0.5.3:GHC.Types.Symbol
                                                          ('Mu.Rpc.SchemaRef
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             QuickstartSchema
                                                             "HelloResponse")))
                                                    ((':)
                                                       @(Mu.Rpc.Method
                                                           @{*}
                                                           ghc-prim-0.5.3:GHC.Types.Symbol
                                                           ghc-prim-0.5.3:GHC.Types.Symbol
                                                           ghc-prim-0.5.3:GHC.Types.Symbol
                                                           (Mu.Rpc.TypeRef
                                                              ghc-prim-0.5.3:GHC.Types.Symbol))
                                                       ('Mu.Rpc.Method
                                                          @{*}
                                                          @ghc-prim-0.5.3:GHC.Types.Symbol
                                                          @ghc-prim-0.5.3:GHC.Types.Symbol
                                                          @ghc-prim-0.5.3:GHC.Types.Symbol
                                                          @(Mu.Rpc.TypeRef
                                                              ghc-prim-0.5.3:GHC.Types.Symbol)
                                                          "SayHi"
                                                          ((':)
                                                             @(Mu.Rpc.Argument
                                                                 @{*}
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol
                                                                 (Mu.Rpc.TypeRef
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol))
                                                             ('Mu.Rpc.ArgSingle
                                                                @{*}
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                @(Mu.Rpc.TypeRef
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol)
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                ('Nothing
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol)
                                                                ('Mu.Rpc.SchemaRef
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   QuickstartSchema
                                                                   "HiRequest"))
                                                             ('[]
                                                                @(Mu.Rpc.Argument
                                                                    @{*}
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol
                                                                    (Mu.Rpc.TypeRef
                                                                       ghc-prim-0.5.3:GHC.Types.Symbol))))
                                                          ('Mu.Rpc.RetStream
                                                             @{*}
                                                             @(Mu.Rpc.TypeRef
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol)
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             ('Mu.Rpc.SchemaRef
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                QuickstartSchema
                                                                "HelloResponse")))
                                                       ((':)
                                                          @(Mu.Rpc.Method
                                                              @{*}
                                                              ghc-prim-0.5.3:GHC.Types.Symbol
                                                              ghc-prim-0.5.3:GHC.Types.Symbol
                                                              ghc-prim-0.5.3:GHC.Types.Symbol
                                                              (Mu.Rpc.TypeRef
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol))
                                                          ('Mu.Rpc.Method
                                                             @{*}
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             @ghc-prim-0.5.3:GHC.Types.Symbol
                                                             @(Mu.Rpc.TypeRef
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol)
                                                             "SayManyHellos"
                                                             ((':)
                                                                @(Mu.Rpc.Argument
                                                                    @{*}
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol
                                                                    (Mu.Rpc.TypeRef
                                                                       ghc-prim-0.5.3:GHC.Types.Symbol))
                                                                ('Mu.Rpc.ArgStream
                                                                   @{*}
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   @(Mu.Rpc.TypeRef
                                                                       ghc-prim-0.5.3:GHC.Types.Symbol)
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   ('Nothing
                                                                      @ghc-prim-0.5.3:GHC.Types.Symbol)
                                                                   ('Mu.Rpc.SchemaRef
                                                                      @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                      @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                      @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                      QuickstartSchema
                                                                      "HelloRequest"))
                                                                ('[]
                                                                   @(Mu.Rpc.Argument
                                                                       @{*}
                                                                       ghc-prim-0.5.3:GHC.Types.Symbol
                                                                       ghc-prim-0.5.3:GHC.Types.Symbol
                                                                       (Mu.Rpc.TypeRef
                                                                          ghc-prim-0.5.3:GHC.Types.Symbol))))
                                                             ('Mu.Rpc.RetStream
                                                                @{*}
                                                                @(Mu.Rpc.TypeRef
                                                                    ghc-prim-0.5.3:GHC.Types.Symbol)
                                                                @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                ('Mu.Rpc.SchemaRef
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   @ghc-prim-0.5.3:GHC.Types.Symbol
                                                                   QuickstartSchema
                                                                   "HelloResponse")))
                                                          ('[]
                                                             @(Mu.Rpc.Method
                                                                 @{*}
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol
                                                                 ghc-prim-0.5.3:GHC.Types.Symbol
...
```

There is more but I got tired of copy-pasting it.  Point is, the error message the user gets when the AnnotatedPackage type family doesn't expand is completely unhelpful. Do you know any technique that would allow me to print a custom `TypeError` message when this happens? I don't know how to do that conditionally on a type family not expanding.

Reply:
Actually, I just found https://kcsongor.github.io/report-stuck-families/ which at first glance seems like it might have a solution.

Reply:
I was going to send you exactly that link, but you were faster!

Reply:
Pushed the custom error messages and documentation. I've never used haddock before, so please let me know if there is anything that I should do differently.